### PR TITLE
Spot instances

### PR DESCRIPTION
**Pull request by whites11**
This is preliminary work to support spot instances on azure opeator.
It's been tested quite a bit, but surely requires more testing from y'all.
I suggest we begin reviewing the PR and merge as soon as we're confident so we can keep testing it as part of the other modifications we'll develop for 14.1.0.

This PR also contains the modifications needed to use latest release of CAPI and CAPZ deps.

WARNING: Because of [this bug](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1151) it is required to edit the `azuremachinepools.exp.infrastructure.cluster.x-k8s.io` CRD to change the type of the `maxPrice` field from `number` to `string` or it is not going to be possible to create spot instance node pools. Regular node pools can be created normally even without the change.

### CODE REVIEW

**Comment by fiunchinho** [comment]
> WARNING: Because of [this bug](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1151) it is required to edit the `azuremachinepools.exp.infrastructure.cluster.x-k8s.io` CRD to change the type of the `maxPrice` field from `number` to `string` or it is not going to be possible to create spot instance node pools. Regular node pools can be created normally even without the change.

do we know if it will take long for upstream to fix it?

File: service/controller/azuremachinepool/handler/nodepool/template/main.json [hunk 1]
```diff
@@ -185,7 +205,8 @@
         "gs-azure-operator.giantswarm.io-version": "[parameters('azureOperatorVersion')]",
         "kubernetes-version": "[parameters('kubernetesVersion')]",
         "min": "[int(parameters('minReplicas'))]",
-        "max": "[int(parameters('maxReplicas'))]"
+        "max": "[int(parameters('maxReplicas'))]",
+        "spot": "[if(parameters('spotInstancesEnabled'), 'true', 'false')]"
```
Comment (fiunchinho) [inline service/controller/azuremachinepool/handler/nodepool/template/main.json:209]: I think this line can be removed

**Review by fiunchinho — CHANGES_REQUESTED** [review]

**Comment by whites11** [comment]
> > WARNING: Because of [this bug](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1151) it is required to edit the `azuremachinepools.exp.infrastructure.cluster.x-k8s.io` CRD to change the type of the `maxPrice` field from `number` to `string` or it is not going to be possible to create spot instance node pools. Regular node pools can be created normally even without the change.
> 
> do we know if it will take long for upstream to fix it?

Adrian just opened the bug and nobody answered yet.
The timeframe I'm expecting is in the order of magnitude of weeks, if we are lucky.

File: service/controller/azuremachinepool/handler/nodepool/template/main.json (hunk 1 above)
Comment (whites11) [reply service/controller/azuremachinepool/handler/nodepool/template/main.json:209]: Which line?

Comment (fiunchinho) [reply service/controller/azuremachinepool/handler/nodepool/template/main.json:209]: oh it's a tag, I thought it was a parameter. Never mind.

**Review by fiunchinho — APPROVED** [review]

**Review by tuommaki — APPROVED** [review]
LGTM

**Pull request merged**